File: analytics/Julia/GetDataFrame.jl
using HDF5
using DataFrames
using Dates

function getDailyDF(f::String) :: DataFrame
    h5f  = h5open(f,"r")
    dd0  = h5f["daily/block0_values"] # Integer64 (datetime)
    dd1  = h5f["daily/block1_values"] # Floats
    dd2  = h5f["daily/block2_values"] # Integers (count and daylight indicator)
    dd0n = h5f["daily/block0_items"]
    dd1n = h5f["daily/block1_items"]
    dd2n = h5f["daily/block2_items"]
    dfd0 = DataFrame(read(dd0)')
    dfd1 = DataFrame(read(dd1)')
    dfd2 = DataFrame(read(dd2)')
    ddn = vcat(read(dd0n), read(dd1n), read(dd2n))

    #+
    # Concatenate dataframes, add column labels and convert datetime
    #-
    dfd = hcat(dfd0,dfd1,dfd2)
    names!(dfd,Symbol.(ddn))
    dfd[:date] = unix2datetime.(div.(dfd[:date],1000000000))

    #+
    # Convert the integer daytime indicator to a categorical, descriptive
    # variable.
    #-
    dfd[:daytime] = CategoricalArray(map(x-> if x==1 return "Day" else return "Night" end , dfd[:daytime]))
    close(h5f)
    return dfd
end

function getHourlyDF(f::String) :: DataFrame
    h5f  = h5open(f,"r")
    dh0 =  h5f["hourly/block0_values"] # Floats dataset
    dh1  = h5f["hourly/block1_values"] # integer dataset (count)
    dh2  = h5f["hourly/block2_values"] # integer64 dataset (time)
    dh0n = h5f["hourly/block0_items"]
    dh1n = h5f["hourly/block1_items"]
    dh2n = h5f["hourly/block2_items"]

    #+
    # Create dataframes from the dataset contents and prepare an
    # array with the field labels
    #
    dfh0 = DataFrame(read(dh0)')
    dfh1 = DataFrame(read(dh1)')
    dfh2 = DataFrame(read(dh2)')
    dfn = vcat(read(dh0n), read(dh1n), read(dh2n))

    #+
    # Concatenate the datasets, assign column names and convert int64 to datetime
    #-
    dfh = hcat(dfh0,dfh1,dfh2)
    names!(dfh,Symbol.(dfn))
    dfh[:time] = unix2datetime.(div.(dfh[:time],1000000000))
    close(h5f)
    return dfh
end

function getRawDF(f::String) :: DataFrame
    h5f  = h5open(f,"r")

    dr0  = h5f["rawdata/block0_values"] # Floats dataset
    dr1  = h5f["rawdata/block1_values"] # integer64 (datetime) dataset
    dr0n = h5f["rawdata/block0_items"]
    dr1n = h5f["rawdata/block1_items"]

    #+
    # Create DataFrames for the block0 (floats) and block1 (integer64 containing
    # times in nanosecods). Note the tick used to transpose the read matrices.
    # Build an array with the field labels from both datasets.
    #-
    dfr0 = DataFrame(read(dr0)')
    dfr1 = DataFrame(read(dr1)')
    drn  = vcat(read(dr1n),read(dr0n))

    #+
    # Concatenate both DataFrames and assign the column names.
    #-
    dfr = hcat(dfr1,dfr0)
    names!(dfr,Symbol.(drn))

    #+
    # Convert the interger64 values to proper times. Remember those
    # times are UTC, but the Julia datatime is plain (has no TZ info)
    #-
    dfr[:time] = unix2datetime.(div.(dfr[:time],1000000000))
    close(h5f)
    return dfr
end

File: analytics/R/GetDataFrame.R
#+
# R Functions to read the weather datasets from the HDF5 files
#-

require(hdf5r)
require(dplyr)
require(nanotime)


#+
# The hdf5r library returns unidimensional datasets as lists and multidimensional
# ones as matrices, which variables in different rows. Those matrices must be transposed
# to be converted to dataframes.
# This function verifies if its argument is a matrix and, if this is the case, it returns
# it transposed.
#-
checkAndTranspose <- function(m) {
  if (is.null(dim(m)))
    return (m)
  else
    return (t(m))
}

#+
# Get the daily averages dataset from an hdf5 file
#-
getDailyDF <- function(file) {
  #+
  # Inner function to convert the daytime indicator from integer (0/1) to 
  # a string value ("Night"/"Day")
  #-
  dayTimeName <- function(x) {
    if (x['daytime']==1)
      return("Day")
    else
      return("Night")
  }
  
  #+
  # Open hdf5 file and data (_values) and labels (_items)
  # datasets.
  #-
  h5f <- H5File$new(filename = file, mode = 'r')
  d0 <- h5f$open('daily/block0_values')
  d1 <- h5f$open('daily/block1_values')
  d2 <- h5f$open('daily/block2_values')
  n0 <- h5f$open('daily/block0_items')
  n1 <- h5f$open('daily/block1_items')
  n2 <- h5f$open('daily/block2_items')
  
  #+
  # Read datasets containing values into matrices
  # (see the comment for checkAndTranspose())
  #-
  a0 <- checkAndTranspose(d0$read()) 
  a1 <- checkAndTranspose(d1$read())
  a2 <- checkAndTranspose(d2$read())
  
  #+
  # Convert the matrices to dataframes and add the
  # column labels
  #-
  df0 <- data.frame(a0)
  colnames(df0) <- n0$read() 
  df1 <- data.frame(a1)
  colnames(df1) <- n1$read() 
  df2 <- data.frame(a2)
  colnames(df2) <- n2$read() 
  
  #+
  # Combine the three dataframes into a global one.
  # Change the daytime column to a factor variable.
  # Change the date value from integer64 to proper time.
  #
  df <- cbind(df0,df1,df2)
  df <- mutate(df, daytime=apply(df,1,dayTimeName)) %>% mutate(daytime=as.factor(daytime)) %>%
        mutate(date=as.POSIXct(nanotime(df$date),tz='UTC'))
  h5f$close()
  return (df)
}
  
#+
# Get the hourly averages dataset from an hdf5 file
#-
getHourlyDF <- function(file) {
  #+
  # Open hdf5 file and data (_values) and labels (_items)
  # datasets.
  #-
  h5f <- H5File$new(filename = file, mode = 'r')
  d0 <- h5f$open('hourly/block0_values')
  d1 <- h5f$open('hourly/block1_values')
  d2 <- h5f$open('hourly/block2_values')
  n0 <- h5f$open('hourly/block0_items')
  n1 <- h5f$open('hourly/block1_items')
  n2 <- h5f$open('hourly/block2_items')
  
  #+
  # Read datasets containing values into matrices
  # (see the comment for checkAndTranspose())
  #-
  a0 <- checkAndTranspose(d0$read()) 
  a1 <- checkAndTranspose(d1$read())
  a2 <- checkAndTranspose(d2$read())
  
  #+
  # Convert the matrices to dataframes and add the
  # column labels
  #-
  df0 <- data.frame(a0)
  colnames(df0) <- n0$read() 
  df1 <- data.frame(a1)
  colnames(df1) <- n1$read() 
  df2 <- data.frame(a2)
  colnames(df2) <- n2$read() 
  
  #+
  # Combine the three dataframes into a global one.
  # Change the date value from integer64 to proper time.
  #
  df <- cbind(df0,df1,df2)
  df <- mutate(df, time=as.POSIXct(nanotime(df$time),tz='UTC'))
  h5f$close()
  return (df)
}

#+
# Get the raw observations dataset from an hdf5 file
#-
getRawDF <- function(file) {
  #+
  # Open hdf5 file and data (_values) and labels (_items)
  # datasets.
  #-
  h5f <- H5File$new(filename = file, mode = 'r')
  d0 <- h5f$open('rawdata/block0_values')
  d1 <- h5f$open('rawdata/block1_values')
  n0 <- h5f$open('rawdata/block0_items')
  n1 <- h5f$open('rawdata/block1_items')

  #+
  # Read datasets containing values into matrices
  # (see the comment for checkAndTranspose())
  #-
  a0 <- checkAndTranspose(d0$read()) 
  a1 <- checkAndTranspose(d1$read())

  #+
  # Convert the matrices to dataframes and add the
  # column labels
  #-
  df0 <- data.frame(a0)
  colnames(df0) <- n0$read() 
  df1 <- data.frame(a1)
  colnames(df1) <- n1$read() 

  #+
  # Combine the two dataframes into a global one.
  # Change the date value from integer64 to proper time.
  #
  df <- cbind(df0,df1)
  df <- mutate(df, time=as.POSIXct(nanotime(df$time),tz='UTC'))
  h5f$close()
  return (df)
}